network/dag/state.go

```diff
@@ -435,72 +429,37 @@ func (s *state) Verify(ctx context.Context) error {
 }
 
 func (s *state) saveEvent(tx stoabs.WriteTx, event Event) error {
-	s.notifiersMux.RLock()
-	defer s.notifiersMux.RUnlock()
-
-	for _, notifier := range s.notifiers {
-		if err := notifier.Save(tx, event); err != nil {
-			return err
-		}
-	}
-
-	return nil
+	var err error
+	s.notifiers.Range(func(_, value any) bool {
+		err := value.(Notifier).Save(tx, event)
+		return err == nil
+	})
+	return err
 }
 
 func (s *state) notify(event Event) {
-	s.notifiersMux.RLock()
-	defer s.notifiersMux.RUnlock()
-
-	for _, notifier := range s.notifiers {
-		notifier.Notify(event)
-	}
+	s.notifiers.Range(func(_, value any) bool {
+		value.(Notifier).Notify(event)
+		return true
+	})
 }
 
-func (s *state) getFailedEvents() []Event {
-	s.notifiersMux.RLock()
-	defer s.notifiersMux.RUnlock()
-
-	failedEvents := make([]Event, 0)
-	for name, notifier := range s.notifiers {
-		events, err := notifier.GetFailedEvents()
+func (s *state) failedEventCount() int {
+	var n int
+	s.notifiers.Range(func(key, value any) bool {
+		events, err := value.(Notifier).GetFailedEvents()
 		if err != nil {
-			log.Logger().Errorf("Failed to retrieve failed events: %v (notifier=%s)", err, name)
-			continue
+			log.Logger().Errorf("failed events lookup %q: %s", key, err)
```

Comment:
our log messages start with a capital, and might be useful to keep the notifier name to aid future analysis. We moved to structured logging recently (for fields worth indexing), so this should probably be something like:
```suggestion
			log.Logger().WithError(err).Errorf("Unable to retrieve failed events for %q", key)
```

Reply:
Even more log bells and whistles. 😢 The core log.Printf is such a beauty.
Any way, "unable to" and "failed to" are kind of useless prefixes in an error message.

Reply:
I can live without the prefix, as long as we keep the field logs (they help log analysis by providing indexable fields). Form (consistency in casing) is probably more of a personal quest; I get annoyed and thus distracted when message formatting/casing isn't consistent. But that might be my own shortcoming..